File: algorithms/267.Palindrome_Permutation_II/Palindrome_Permutation_II.cc
<<<<<<< HEAD
class Solution {
public:
    vector<string> generatePalindromes(string s) {
        vector<string> palindromes;
        unordered_map<char, int> counts;
        for (char c : s) counts[c]++;
        int odd = 0; char mid; string half;
        for (auto p : counts) {
            if (p.second & 1) {
                odd++, mid = p.first;
                if (odd > 1) return palindromes;
            }
            half += string(p.second / 2, p.first);
        }
        palindromes = permutations(half);
        for (string& p : palindromes) {
            string t(p);
            reverse(t.begin(), t.end());
            if (odd) t = mid + t;
            p += t;
        }
        return palindromes;
    }
private:
    vector<string> permutations(string& s) {
        vector<string> perms;
        string t(s);
        do {
            perms.push_back(s);
            next_permutation(s.begin(), s.end());
        } while (s != t);
        return perms;
    }
=======
class Solution {
public:
    vector<string> generatePalindromes(string s) {
        vector<string> palindromes;
        unordered_map<char, int> counts;
        for (char c : s) counts[c]++;
        int odd = 0; char mid; string half;
        for (auto p : counts) {
            if (p.second & 1) {
                odd++, mid = p.first;
                if (odd > 1) return palindromes;
            }
            half += string(p.second / 2, p.first);
        }
        palindromes = permutations(half);
        for (string& p : palindromes) {
            string t(p);
            reverse(t.begin(), t.end());
            if (odd) t = mid + t;
            p += t;
        }
        return palindromes;
    }
private:
    vector<string> permutations(string& s) {
        vector<string> perms;
        string t(s);
        do {
            perms.push_back(s);
            next_permutation(s.begin(), s.end());
        } while (s != t);
        return perms;
    }
>>>>>>> 6200c8704614e918c8bfa5357c648dd1b4f7eb74
};

File: algorithms/267.Palindrome_Permutation_II/Palindrome_Permutation_II.py
<<<<<<< HEAD
class Solution(object):
    def generatePalindromes(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        # check element
        d = {}
        for i in s:
            if i not in d:
                d[i] = 1
            else:
                d[i] += 1
        oddcount = 0
        oddchar = ''
        for char in d:
            if d[char] & 1:
                oddcount += 1
                d[char] -= 1
                oddchar = char
            d[char] //= 2
        if oddcount > 1:
            return []
            
        # do permutation to iterate through all possible str
        # note that we just need to permute the first half of the
        # palindrome
        p = ''
        for char in d:
            p += char * d[char]
        if not p:
            return [oddchar]
        p = sorted(list(p))
        l = len(p)
        res = []
        while 1:
            res.append(''.join(p)+oddchar+''.join(p[::-1]))
            n = l-1
            while n > 0 and p[n - 1] >= p[n]:
                n -= 1
            if n == 0:
                break
            t = n
            while t < l and p[n - 1] < p[t]:
                t += 1
            temp = p[t-1]
            p[t-1] = p[n-1]
            p[n-1] = temp
            p[n:] = p[n:][::-1]
        return res
=======
class Solution(object):
    def generatePalindromes(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        # check element
        d = {}
        for i in s:
            if i not in d:
                d[i] = 1
            else:
                d[i] += 1
        oddcount = 0
        oddchar = ''
        for char in d:
            if d[char] & 1:
                oddcount += 1
                d[char] -= 1
                oddchar = char
            d[char] //= 2
        if oddcount > 1:
            return []
            
        # do permutation to iterate through all possible str
        # note that we just need to permute the first half of the
        # palindrome
        p = ''
        for char in d:
            p += char * d[char]
        if not p:
            return [oddchar]
        p = sorted(list(p))
        l = len(p)
        res = []
        while 1:
            res.append(''.join(p)+oddchar+''.join(p[::-1]))
            n = l-1
            while n > 0 and p[n - 1] >= p[n]:
                n -= 1
            if n == 0:
                break
            t = n
            while t < l and p[n - 1] < p[t]:
                t += 1
            temp = p[t-1]
            p[t-1] = p[n-1]
            p[n-1] = temp
            p[n:] = p[n:][::-1]
        return res
>>>>>>> 6200c8704614e918c8bfa5357c648dd1b4f7eb74
